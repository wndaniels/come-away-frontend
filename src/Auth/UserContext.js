import React from "react";
/**
 *  Context:
 *
 * Provides currentUser object and setter for it throughout app.
 */
const UserContext = React.createContext();

export default UserContext;
