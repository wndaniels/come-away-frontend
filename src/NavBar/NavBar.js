import { useContext, useEffect, useState } from "react";
import { Link, NavLink } from "react-router-dom";
import UserContext from "../Auth/UserContext.js";
import ComeAwayApi from "../Api/api.js";
import useLocalStorage from "../hooks/useLocalStorage.js";
import jwt from "jsonwebtoken";

export const TOKEN_STORAGE_ID = "comeaway-token";

const NavBar = () => {
  const [token, setToken] = useLocalStorage(TOKEN_STORAGE_ID);
  const [calUserId, setCalUserId] = useState([]);
  const [infoLoaded, setInfoLoaded] = useState(false);
  const [formError, setFormError] = useState([]);

  const { currentUser, setCurrentUser, logout } = useContext(UserContext);

  useEffect(
    function loadUserInfo() {
      async function getCurrentUser() {
        if (token) {
          try {
            let { username } = jwt.decode(token);
            ComeAwayApi.token = token;
            let currentUser = await ComeAwayApi.getCurrentUser(username);
            setCurrentUser(currentUser);
          } catch (errors) {
            setFormError(errors);
            setCurrentUser(null);
          }
        }
        setInfoLoaded(true);
      }
      setInfoLoaded(false);
      getCurrentUser();
    },
    [token, setCurrentUser]
  );

  useEffect(() => {
    async function getCalDataByUser() {
      const userCalData = await ComeAwayApi.getAllCals();
      if (currentUser)
        userCalData.forEach((d) => {
          try {
            if (currentUser.id === d.userId) {
              setCalUserId(d.userId);
            }
          } catch (errors) {
            return null;
          }
        });
    }

    getCalDataByUser();
  }, [currentUser]);

  const loggedOutUser = () => {
    return (
      <ul className="navbar-nav ml-auto">
        <li className="nav-item mr-4">
          <NavLink className="nav-link" to="/login">
            Login
          </NavLink>
        </li>
        <li className="nav-item mr-4">
          <NavLink className="nav-link" to="/signup">
            Sign Up
          </NavLink>
        </li>
      </ul>
    );
  };

  const loggedInUser = () => {
    if (currentUser.id !== calUserId) {
      return (
        <ul className="navbar-nav ml-auto">
          <li className="nav-item mr-4">
            <NavLink
              className="nav-link"
              to={`/${currentUser.username}/calendar/setup/1`}
            >
              Create Calendar
            </NavLink>
          </li>
          <li className="nav-item mr-4">
            <NavLink
              className="nav-link"
              to={`/${currentUser.username}/profile/edit`}
            >
              Profile
            </NavLink>
          </li>
          <li className="nav-item mr-4">
            <NavLink className="nav-link" onClick={logout}>
              Logout
            </NavLink>
          </li>
        </ul>
      );
    } else {
      return (
        <ul className="navbar-nav ml-auto">
          <li className="nav-item mr-4">
            <NavLink
              className="nav-link"
              to={`/${currentUser.username}/calendar/edit`}
            >
              Edit Calendar
            </NavLink>
          </li>
          <li className="nav-item mr-4">
            <NavLink
              className="nav-link"
              to={`/${currentUser.username}/profile/edit`}
            >
              Profile
            </NavLink>
          </li>
          <li className="nav-item mr-4">
            <NavLink className="nav-link" onClick={logout}>
              Logout
            </NavLink>
          </li>
        </ul>
      );
    }
  };

  return (
    <nav className="NavBar navbar navbar-expand-sm">
      <div className="container-fluid">
        <Link className="navbar-brand" to="/">
          COME/AWAY
        </Link>
        {currentUser ? loggedInUser() : loggedOutUser()}
      </div>
    </nav>
  );
};

export default NavBar;
