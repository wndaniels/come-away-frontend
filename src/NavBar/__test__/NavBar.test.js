import "../../matchMedia.mock";
import React from "react";
import { getByRole, render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import "@testing-library/jest-dom";
import NavBar from "../NavBar";
import UserContext from "../../Auth/UserContext";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
}));

describe("NavBar", () => {
  it("renders without crashing", () => {
    const { asFragment } = render(
      <Wrapper>
        <NavBar />
      </Wrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("can find heading", () => {
    const { getByRole } = render(
      <Wrapper>
        <NavBar />
      </Wrapper>
    );

    const anchorTag = getByRole("link", {
      name: "COME/AWAY",
      class: "navbar-brand",
    });
    expect(anchorTag).toBeInTheDocument();
  });
});

function Wrapper({ children }) {
  return (
    <UserContext.Provider
      value={{
        currentUser: jest.fn(),
        setCurrentUser: jest.fn(),
        login: jest.fn(),
        signup: jest.fn(),
        logout: jest.fn(),
      }}
    >
      {children}
    </UserContext.Provider>
  );
}
