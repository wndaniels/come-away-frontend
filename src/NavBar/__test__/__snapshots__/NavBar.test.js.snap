// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar renders without crashing 1`] = `
<DocumentFragment>
  <nav
    class="NavBar navbar navbar-expand-sm"
  >
    <div
      class="container-fluid"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        COME/AWAY
      </a>
      <ul
        class="navbar-nav ml-auto"
      >
        <li
          class="nav-item mr-4"
        >
          <a
            class="nav-link"
            href="/undefined/calendar/setup/1"
          >
            Create Calendar
          </a>
        </li>
        <li
          class="nav-item mr-4"
        >
          <a
            class="nav-link"
            href="/undefined/profile/edit"
          >
            Profile
          </a>
        </li>
        <li
          class="nav-item mr-4"
        >
          <a
            class="nav-link"
            href="/"
          >
            Logout
          </a>
        </li>
      </ul>
    </div>
  </nav>
</DocumentFragment>
`;
